File: java/leetcode/editor/cn/MaxConsecutiveOnesIii.java
package leetcode.editor.cn;
		  
import java.util.*;
import leetcode.editor.cn.struct.*;
		  
//[1004] 最大连续1的个数 III
/* [1,1,1,0,0,0,1,1,1,1,0],2 */
public class MaxConsecutiveOnesIii{
    public static void main(String[] args) {
        Solution solution = new MaxConsecutiveOnesIii().new Solution();
    }
    
//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int longestOnes(int[] nums, int k) {
        int n = nums.length;
        int res = 0;
        int l = 0, r = 0;
        int zeros = 0;
        while (r < n) {
            if (nums[r] == 0)
                zeros ++;
            while (zeros > k) {
                if (nums[l++] == 0)
                    zeros --;
            }
            res = Math.max(res, r - l + 1);
            r ++;
        }
        return res;

    }
}
//leetcode submit region end(Prohibit modification and deletion)

}

File: src/main/java/leetcode/editor/cn/MaxConsecutiveOnesIii.kt
package leetcode.editor.cn
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import util.to1DIntArray


// 1004 最大连续1的个数 III
/* [1,1,1,0,0,0,1,1,1,1,0],2 */

fun main() {
    val result = MaxConsecutiveOnesIii.Solution()::longestOnes
    assertThat(result("[1,1,1,0,0,0,1,1,1,1,0]".to1DIntArray(),2), equalTo(6 ))
    assertThat(result("[0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1]".to1DIntArray(),3), equalTo(10))
}

class MaxConsecutiveOnesIii{
//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    fun longestOnes(A: IntArray, K: Int): Int {
        val n = A.size
        var left = 0
        var lsum = 0
        var rsum = 0
        var ans = 0
        for (right in 0 until n) {
            rsum += 1 - A[right]
            while (lsum < rsum - K) {
                lsum += 1 - A[left]
                ++left
            }
            ans = Math.max(ans, right - left + 1)
        }
        return ans

    }
}
//leetcode submit region end(Prohibit modification and deletion)

}